ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.utils.io
+
+import io.ktor.utils.io.internal.*
+import io.ktor.utils.io.locks.*
+import kotlinx.atomicfu.*
+import kotlinx.io.*
+import kotlin.concurrent.*
+
+@InternalAPI
+public val CHANNEL_MAX_SIZE: Int = 4 * 1024
+
+/**
+ * Sequential (non-concurrent) byte channel implementation
+ */
+public class ByteChannel(public val autoFlush: Boolean = false) : ByteReadChannel, BufferedByteWriteChannel {
+    private val _closedCause = atomic<CloseToken?>(null)
+    private val slot = AwaitingSlot()
+    private val flushBuffer: Buffer = Buffer()
```

Comment:
Is the intent with this class to eventually drop the three buffers and flush orchestration in favour of just using a single buffer with some synchronization?  Seems like we could just have a single bounded buffer that blocks writes until there's space.

Reply:
The problem is that not all operations writing this buffer are suspend

Reply:
Anyway let's discuss this idea